### ISSUE TITLE
'module restore' collection order

### ISSUE THREAD

**Issue by kmanalo** [TACC/Lmod#148]
I am trying to reproduce MODULEPATH ordering behavior when restoring from a collection, as it pertains to combined use of `module use` and prepending to MODULEPATH.  I have distilled it into the example below.

Here are the contents of three files

`/data/share/A/modulefiles/mysoft/17.0.0.lua`

```
LmodMessage("Hello from folder A")
```

`/data/share/B/modulefiles/mysoft/17.0.0.lua`

```
LmodMessage("Hello from folder B")
```

`/usr/local/modulefiles/Core/test/1.0.0.lua`

```
prepend_path("MODULEPATH", "/data/share/B/modulefiles")
```

My default modules look like this

```
$ ml av

--------------- /usr/local/modulefiles/Core ----------------
   intel/16.0.3    python/3.5    test/1.0.0

---------- /usr/local/lmod/lmod/modulefiles/Core -----------
   lmod/6.4.1    settarg/6.4.1

Use "module spider" to find all possible modules.
Use "module keyword key1 key2 ..." to search for all
possible modules matching any of the "keys".

```
1. I load the test module for access to 'mysoft' in A, then perform a module use for the B folder. MODULEPATH goes in order A,B,...
   
   ```
   $ module load test
   $ module use /data/share/A/modulefiles 
   $ module load mysoft
   Hello from folder A
   $ module avail
   
   ---------------- /data/share/A/modulefiles -----------------
   mysoft/17.0.0 (D)
   
   ---------------- /data/share/B/modulefiles -----------------
   mysoft/17.0.0
   
   --------------- /usr/local/modulefiles/Core ----------------
   intel/16.0.3    python/3.5    test/1.0.0 (L)
   
   ---------- /usr/local/lmod/lmod/modulefiles/Core -----------
   lmod/6.4.1    settarg/6.4.1
   
   Where:
   L:  Module is loaded
   D:  Default Module
   
   Use "module spider" to find all possible modules.
   Use "module keyword key1 key2 ..." to search for all
   possible modules matching any of the "keys".
   ```
2. I save the collection to 'm', and restore it, but MODULEPATH folders go in order B,A,...

```
module save m
module restore m # in same shell or new shell will result in 
Hello from folder B
---------------- /data/share/B/modulefiles -----------------
   mysoft/17.0.0 (L,D)

---------------- /data/share/A/modulefiles -----------------
   mysoft/17.0.0

--------------- /usr/local/modulefiles/Core ----------------
   intel/16.0.3    python/3.5    test/1.0.0 (L)

---------- /usr/local/lmod/lmod/modulefiles/Core -----------
   lmod/6.4.1    settarg/6.4.1

  Where:
   L:  Module is loaded
   D:  Default Module

Use "module spider" to find all possible modules.
Use "module keyword key1 key2 ..." to search for all
possible modules matching any of the "keys".

```

While likely not a problem (I used the same software and version artificially), is this the intended behavior? Or should the MODULEPATH order stay the same at the point of saving?

**Comment by rtmclay** [comment]
This looks like a bug.  Please send do the following:

       module -D save m 2> save.log
       module -D restore m 2> restore.log

and send me in a tar ball: save.log, restore.log and ~/.lmod.d/m

Thanks,

**Comment by rtmclay** [comment]
I have looked at this and this is what I intended.  The way that a restore works is:

1. (Force) Purge all modules
2. Set MODULEPATH to the MODULEPATH stored in the collection
3. Load the modules in the collection in the order they are listed

What Lmod doesn't do is reset the module path to the MODULEPATH in the collection by redoing step 2. This would force the reloading of "mysoft" from the "B" version to the "A" version.    This is a reasonable request but it may be a while before I can get to this.  These actions happen in src/MT.lua in the getMTfromFile() routine if you wish to look at it and submit a PR.

The trick here will be to figure which paths to "prepend_path" on MODULEPATH and avoid a load module storm where Lmod is loading and unloading the same modulefiles.
